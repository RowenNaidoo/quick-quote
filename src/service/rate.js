import 'whatwg-fetch'

export const getSpotRate = async (fromCurrency, toCurrency, amount) => (
  window.fetch(`https://api.ofx.com/PublicSite.ApiService/OFX/spotrate/Individual/${fromCurrency}/${toCurrency}/${amount}?format=json`)
    .then(response => response.json())
)