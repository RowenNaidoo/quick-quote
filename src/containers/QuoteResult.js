import React, { useState } from 'react'
import styled from 'styled-components'

import { getSpotRate } from '../service/rate'
import Button from '../components/Button'

const QuoteResult = ({ match: { params }, history }) => {
  const { fromCurrency, toCurrency, amount } = params
  const [quoteData, setQuoteData] = useState({})

  React.useEffect(() => {
    let cancel = false

    const runEffect = async () => {
      const quoteResponse = await getSpotRate(fromCurrency, toCurrency, amount)
      if (cancel) { return }
      setQuoteData(quoteResponse)
    }
    runEffect()

    return () => {
      cancel = true
    }
  }, [])

  const startNewQuote = () => {
    history.push('/')
  }

  return (
    <Container>
      <MainHeader>
        OFX Customer Rate
      </MainHeader>
      <Rate id='rate'>
        {quoteData.CustomerRate}
      </Rate>
      <AmountContainer>
        <Header>
          From
      </Header>
        <Currency>
          {fromCurrency}
        </Currency>
        <Amount id='fromAmount'>
          {parseFloat(amount).toFixed(2)}
        </Amount>
      </AmountContainer>
      <AmountContainer>
        <Header>
          To
        </Header>
        <Currency>
          {toCurrency}
        </Currency>
        <Amount id='toAmount'>
          {parseFloat(quoteData.CustomerAmount).toFixed(2)}
        </Amount>
      </AmountContainer>
      <ButtonContainer>
        <Button buttonText='START NEW QUOTE' onButtonClicked={startNewQuote} />
      </ButtonContainer>
    </Container>
  )
}

const Container = styled.div`
  width: 250px;
  display: flex;
  flex-direction: column;
  margin: auto;
`
const MainHeader = styled.h1`
  font-size: 24px;
  font-weight:normal;
  margin-bottom: 5px;
  text-align: center;
`
const Header = styled.h1`
  font-size: 24px;
  font-weight:normal;
  margin-bottom: 5px;
`
const Rate = styled.div`
  color:green;
  text-align: center;
  font-size:  48px;
`
const AmountContainer = styled.div`
  padding-left: 30px;
  font-size: 35px;
`
const Currency = styled.span`
  font-size: 35px;
  margin-right: 5px;
`
const Amount = styled.span`
  font-size: 35px;
  color: blue;
`
const ButtonContainer = styled.div`
  text-align: center;
  margin-top: 20px;
`

export default QuoteResult