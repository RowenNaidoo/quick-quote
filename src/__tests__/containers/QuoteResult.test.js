import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import QuoteResult from '../../containers/QuoteResult'
import * as rateService from '../../service/rate'


describe('<App />', () => {
  let useEffect
  let component

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  }

  const spotratePromise = new Promise((resolve) => {
    resolve({
      CustomerRate: 0.6209,
      CustomerAmount: 62.09
    })
  })

  const defaultProps = {
    match: { params: { fromCurrency: 'AUD', toCurrency: 'USD', amount: 100 } }
  }

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect')
    mockUseEffect()
    mockUseEffect()

    rateService.getSpotRate = jest.fn().mockReturnValue(spotratePromise)
    component = shallow(<QuoteResult {...defaultProps} />)
  })

  test('matches snapshot', () => {
    component = renderer.create(<QuoteResult {...defaultProps} />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('displays correct rates and values', () => {
    expect(component.find('#rate').text()).toEqual('0.6209')
    expect(component.find('#fromAmount').text()).toEqual(parseFloat('100').toFixed(2))
    expect(component.find('#toAmount').text()).toEqual(parseFloat('62.09').toFixed(2))
  })
})