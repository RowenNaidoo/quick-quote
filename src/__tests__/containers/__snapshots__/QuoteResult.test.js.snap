// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div
  className="sc-AxirZ hqQzZQ"
>
  <h1
    className="sc-AxiKw kmAhWr"
  >
    OFX Customer Rate
  </h1>
  <div
    className="sc-AxhUy eqSEOB"
    id="rate"
  />
  <div
    className="sc-AxgMl igjTzG"
  >
    <h1
      className="sc-AxhCb jWLGde"
    >
      From
    </h1>
    <span
      className="sc-AxheI kQEMLG"
    >
      AUD
    </span>
    <span
      className="sc-Axmtr jfobaw"
      id="fromAmount"
    >
      100.00
    </span>
  </div>
  <div
    className="sc-AxgMl igjTzG"
  >
    <h1
      className="sc-AxhCb jWLGde"
    >
      To
    </h1>
    <span
      className="sc-AxheI kQEMLG"
    >
      USD
    </span>
    <span
      className="sc-Axmtr jfobaw"
      id="toAmount"
    >
      NaN
    </span>
  </div>
  <div
    className="sc-AxmLO kQhVGH"
  >
    <button
      className="sc-AxjAm bJArkW"
      onClick={[Function]}
    >
      START NEW QUOTE
    </button>
  </div>
</div>
`;
